iup/__init__.py

```diff
@@ -0,0 +1,358 @@
+        self = (
+            self.select(["region", "date", "estimate"])
+            .sort("date")
+            .with_columns(estimate=pl.col("estimate").cum_sum().over("region"))
+            .join(last_cumulative, on="region")
+            .with_columns(estimate=pl.col("estimate") + pl.col("last_cumulative"))
+            .drop("last_cumulative")
+        )
+
+        self = CumulativeUptakeData(self)
+
+        return self
+
+
+class CumulativeUptakeData(UptakeData):
+    def validate(self):
+        self.assert_columns_found(["region", "date", "estimate"])
+        self.assert_columns_type(["date"], pl.Date)
+        self.assert_columns_type(["region"], pl.Utf8)
+        self.assert_columns_type(["estimate"], pl.Float64)
+        assert (
+            self["estimate"] >= 0.0
+        ).all(), "Cumulative uptake estimates cannot be negative."
+
+    def to_incident(self) -> IncidentUptakeData:
+        self = (
+            self.with_columns(
+                season=(
+                    pl.col("date").dt.year()
+                    + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0)
+                ).cast(pl.Utf8),
+                estimate=pl.col("estimate").diff().over("region"),
+                elapsed=(pl.col("date") - pl.col("date").first())
+                .over("region")
+                .dt.total_days()
+                .cast(pl.Float64),
+            )
+            .with_columns(
+                estimate=pl.col("estimate")
+                .fill_null(pl.col("estimate").first())
+                .over("region"),
+                interval=pl.col("elapsed").diff().over("region"),
+            )
+            .with_columns(
+                interval=pl.col("interval")
+                .fill_null(pl.col("elapsed").first())
+                .over("region")
+            )
+            .with_columns(estimate=(pl.col("estimate") / pl.col("interval")))
+            .with_columns(previous=pl.col("estimate").shift(1).over("region"))
+        )
+
+        self = IncidentUptakeData(self)
+
+        return self
+
+    @staticmethod
+    def split_train_test(uptake_data_list, start_date, side):
+        if side == "train":
+            out = (
+                pl.concat(uptake_data_list)
+                .sort("date")
+                .filter(pl.col("date") < start_date)
+            )
+        if side == "test":
+            out = (
+                pl.concat(uptake_data_list)
+                .sort("date")
+                .filter(pl.col("date") >= start_date)
+            )
+
+        out = CumulativeUptakeData(out)
+
+        return out
+
+
+def get_nis(
+    data_path,
+    region_col,
+    date_col,
+    estimate_col,
+    rollout,
+    filters=None,
+) -> CumulativeUptakeData:
+    out = (
+        (pl.read_csv(data_path))
+        .with_columns(estimate=pl.col(estimate_col).cast(pl.Float64, strict=False))
+        .drop_nulls(subset=["estimate"])
+    )
+
+    if filters is not None:
```

Comment:
This should be its own separate function, that deserves some unit testing!

Reply:
Issue #33